[AIRFLOW-5073] Change SQLSensor to keep poking receving NULL, '0'

uncletoxa opened the pull request:
This PR is targeting Airflow 2.0+ as discussed [before](https://github.com/apache/airflow/pull/5688).

### Jira

- [x] My PR addresses the following:
  - https://issues.apache.org/jira/browse/AIRFLOW-5073


### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:
- switches NULL response to keep poking instead of 'success'
- removes `allow_true`  parameter to decrease params clutter.

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
- tests if Sensor returns False on `None` response from DB
- removes testing of `allow_null` parameter

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### Code Quality

- [X] Passes `flake8`

uncletoxa committed:
Change SQLSensor to treat NULL as keep poking. Remove allow_null
parameter to decrease clutter.

uncletoxa committed:
Merge branch 'master' into master

ashb commented on airflow/sensors/sql_sensor.py:
```diff
@@ -61,14 +58,13 @@ class SqlSensor(BaseSensorOperator):
 
     @apply_defaults
     def __init__(self, conn_id, sql, parameters=None, success=None, failure=None, fail_on_empty=False,
-                 allow_null=True, *args, **kwargs):
```
I would say leave this as an option but set just change the default please.

uncletoxa replied:
@ashb I strongly disagree. Why would we leave this parameter? First of all, I think that majority of users would never use this option if NULL is 'keep poking' result for sensor. For those who want to leave this weird sensor behavior there is still an option via 'success' parameter -- it is described in changelist.
Long story short: I don't think we should support this bug in new version where we can break compatibility.

ashb commented on UPDATING.md:
```diff
@@ -39,6 +39,13 @@ assists users migrating to a new version.
+
+SQLSensor now keeps poking after receiving `NULL`. Previously `NULL` response
+has been treated as success criteria. Previous behaviour is still achievable
+setting param `success` to `lambda x: str(x) == 'None'`.
```
```suggestion
setting param `success` to `lambda x: x is None or str(x) not in ('0', '')`.
```

ashb commented on airflow/sensors/sql_sensor.py:
```diff
@@ -102,6 +98,4 @@ def poke(self, context):
                 raise AirflowException("self.success is present, but not callable -> {}".format(self.success))
-        if self.allow_null:
-            return str(first_cell) not in ('0', '')
         return str(first_cell) not in ('0', '', 'None')
```
Waaaait. I just noticed. This means if I do `INSERT VALUES("None")` in my table it would fail. That is a bug in the original PR that we need to fix first.

ashb replied:
(By first I mean as a separate PR that we can apply to the 1.10.x release)

uncletoxa replied:
I can't figure out how to fix already merged pull request. (btw fixed in current PR)

ashb replied:
There's no way to change the merged PR. You'd have to create a new PR with that change please.

uncletoxa replied:
I'm confused a little bit. That PR should be from `v1-10-stable` branch, right?

ashb replied:
No, just a separate PR agsint the master branch (as we need the change against  this line - `str(first_cell) not in ('0', '', 'None')`) Once that is in master I will cherry-pick it back to the release branch.

uncletoxa replied:
fixed here https://github.com/apache/airflow/pull/6336

uncletoxa committed:
Update UPDATING.md

Co-Authored-By: Ash Berlin-Taylor <ash_github@firemirror.com>

uncletoxa committed:
[AIRFLOW-5073] Treat NULL as keep poking in SQL sensor

Change SQL sensor to exclude NULL result from success criteria. Remove
allow_null parameter from sensor to avoid extra clutter.

uncletoxa committed:
[AIRFLOW-5073] Optionally treat NULL as keep poking in SQLSensor

Change SQLSensor to treat NULL as fail. Add boolean allow_null
parameter to support legacy behavior passing NULL as success.

uncletoxa committed:
[AIRFLOW-5073] Optionally treat NULL as keep poking in SQLSensor

Change SQLSensor to treat NULL as fail. Remove allow_null
parameter to decrease clutter.

uncletoxa committed:
Change SQLSensor to treat NULL as keep poking.

Remove allow_null parameter to decrease clutter.

uncletoxa committed:
Change SQLSensor to treat NULL as keep poking.

Remove allow_null parameter to decrease clutter.

uncletoxa committed:
Merge branch 'master' into master

uncletoxa committed:
Fix tests, provide correct UPDATING.MD

uncletoxa committed:
Merge branch 'master' into master

uncletoxa committed:
Merge branch 'master' into master

ashb commented on tests/sensors/test_sql_sensor.py:
```diff
@@ -103,6 +103,18 @@ def test_sql_sensor_postgres_poke(self, mock_hook):
         mock_get_records.return_value = []
         self.assertFalse(t.poke(None))
 
+        mock_get_records.return_value = [['None']]
```
This is when it returns the string `"None"` ritght?

Shouldn't we also have a case where we test the behaviour with `[[None]]` as the return?

uncletoxa replied:
I guess `mock_get_records.return_value = []` would check exactly that issue. But I added another case to test for `[[None]]` explicitly

ashb reviewed (commented):
(Sorry for taking so long)

uncletoxa committed:
add testing for None

uncletoxa committed:
Merge branch 'master' of https://github.com/ToxaZ/airflow

ashb commented on UPDATING.md:
```diff
@@ -41,6 +41,13 @@ assists users migrating to a new version.
 
+### Changes to SQLSensor
+
+SQLSensor now consistent with python `bool()` function. It will resolve after receiving any value 
```
This needs to explicitly mention that then `allow_null` parameter has been removed.

ashb reviewed (commented):
LGTM once you make the small change to the UPDATING.md file

(Sorry it's taken so long and lots of back-and-forth for a tiny change)

uncletoxa committed:
Update UPDATING.md

ashb committed:
Update UPDATING.md
